### PR TITLE
Mark ~all LSP methods as `const`.

### PR DESCRIPTION
**Pull request by jvilk-stripe**
Mark ~all LSP methods as `const`. This change moves all `LSPLoop` mutations to a few core functions.

Previously, `handleSorbetWorkspaceEdit`, `tryFastPath`, and `runSlowPath` would each mutate a bunch of things in LSPLoop as they processed an incoming edit:

* `handleSorbetWorkspaceEdit` would update the set of files currently open in the editor.
* `tryFastPath`/`runSlowPath` would update:
  * Indexed ASTs.
  * The contents of files stored in `initialGS`.
  * Global state hashes (used to determine if update takes fast path)

Now, all of these updates are encapsulated in a `FileUpdates` struct that is contained on the output of a typechecker run. A non-const method, `commitTypecheckRun`, commits those updates to `LSPLoop` and pushes diagnostics out to the editor.

In addition, I have removed some action-at-a-distance from LSP queries. Previously, `runLSPQuery` mutated `initialGS` because `tryFastPath` could technically take a slow path and clone `initialGS` before typechecking. This mutation is incompatible with `const`. Thus, I have removed this method, and changed `tryFastPath`/`runSlowPath` to accept an optional `Query` parameter.

### Motivation

This is a necessary step toward making the "slow path" cancelable. A subsequent PR will add in the ability to abort a running slow path, which will skip committing the updates and re-enqueue the un-committed edits for processing. 

### Test plan

All existing tests pass. I added a new test to cover a new corner case that has arisen now that `handleSorbetWorkspaceEdit` no longer mutates `openFiles`.

### CODE REVIEW

File: main/lsp/lsp.h [hunk 1]
```diff
@@ -169,18 +180,22 @@ class LSPLoop {
         std::vector<std::unique_ptr<core::lsp::QueryResponse>> responses;
         // The global state, post-typechecking.
         std::unique_ptr<core::GlobalState> gs;
+        // The edit applied to `gs`.
+        std::optional<LSPLoop::FileUpdates> updates;
         bool tookFastPath;
     };
     /** Conservatively rerun entire pipeline without caching any trees */
-    TypecheckRun runSlowPath();
+    TypecheckRun runSlowPath(std::optional<FileUpdates> updates,
```
Comment (DarkDimius) [inline main/lsp/lsp.h]: does it have to take `optional`? can it just take `FileUpdates` where all vectors are empty?

Comment (DarkDimius) [reply main/lsp/lsp.h]: (ditto in TypecheckRun)

File: main/lsp/requests/sorbet_workspace_edit.cc [hunk 2]
```diff
@@ -19,29 +19,29 @@ string readFile(string_view path, const FileSystem &fs) {
...
+                                     const core::GlobalState &initialGS, string_view path) {
     if (updates.find(path) != updates.end()) {
-        return updates[path];
+        return updates[path].contents;
     }
 
-    auto currentFileRef = initialGS->findFileByPath(path);
+    auto currentFileRef = initialGS.findFileByPath(path);
     if (currentFileRef.exists()) {
-        return string(currentFileRef.data(*initialGS).source());
```
Comment (DarkDimius) [inline main/lsp/requests/sorbet_workspace_edit.cc]: nice!

File: main/lsp/lsp.h [hunk 3]
```diff
@@ -63,18 +63,30 @@ class LSPLoop {
         CounterState counters;
     };
 
+    /**
+     * Encapsulates an update to LSP's file state.
+     */
+    struct FileUpdates {
+        std::vector<std::shared_ptr<core::File>> updatedFiles;
+        // Note: `string_view`s point to `string`s owned by `core::File`s in `updatedFiles`
+        std::vector<std::string_view> openedFiles;
```
Comment (jvilk-stripe) [inline main/lsp/lsp.h]: Note: I identify files with `string_view` because they might be new files, so they do not have a valid ID in `initialGS` yet.

Comment (DarkDimius) [reply main/lsp/lsp.h]: who owns the strings that those string_views come from? I think this is potential footgun. Could we make it store full `string`s instead?

File: main/lsp/updates.cc [hunk 4]
```diff
@@ -219,58 +282,51 @@ bool LSPLoop::canTakeFastPath(const vector<shared_ptr<core::File>> &changedFiles
...
+            updates.updatedFileHashes.push_back(make_pair(f->path(), hashes[i]));
         }
+        core::NameHash::sortAndDedupe(changedHashes);
     }
 
     if (takeFastPath) {
         Timer timeit(logger, "fast_path");
-        // Drop any indexing errors produced during `updateFile`, as we are going to run indexing a second time with
-        // finalGS. (Note: Flushing is disabled in LSP mode, so we have to drain.)
-        errorQueue->drainWithQueryResponses();
```
Comment (jvilk-stripe) [inline main/lsp/updates.cc]: No longer needed now that we run `updateFile` on `initialGS` at the *end* of the typechecking run -- in `commitTypecheckRun`.

File: main/lsp/updates.cc [hunk 5]
```diff
@@ -293,10 +349,12 @@ LSPLoop::TypecheckRun LSPLoop::tryFastPath(unique_ptr<core::GlobalState> gs,
         ENFORCE(initialGS->errorQueue->isEmpty());
         vector<ast::ParsedFile> updatedIndexed;
         for (auto &f : subset) {
+            unique_ptr<KeyValueStore> kvstore; // nullptr
```
Comment (jvilk-stripe) [inline main/lsp/updates.cc]: LSPLoop has a `kvstore`, but I can't use it here as `indexOne` wants a non-const reference. Thankfully, LSPLoop's `kvstore` is always `null`.

File: main/lsp/lsp.h (hunk 3 above)
Comment (jvilk-stripe) [reply main/lsp/lsp.h]: @DarkDimius see `// Note:` in source code.

File: main/lsp/updates.cc (hunk 5 above)
Comment (DarkDimius) [reply main/lsp/updates.cc]: can we add an ENFORCE to check the `thankfully` clause ? :-)

File: main/lsp/lsp.h (hunk 3 above)
Comment (jvilk-stripe) [reply main/lsp/lsp.h]: If it still makes you nervous, I can make them `string`s.

File: main/lsp/updates.cc (hunk 5 above)
Comment (jvilk-stripe) [reply main/lsp/updates.cc]: (If we want to use an actual `kvstore` in future, we could "cheat" by making it an argument to the function.)

File: main/lsp/lsp.h (hunk 3 above)
Comment (DarkDimius) [reply main/lsp/lsp.h]: can those vectors be vectors of `shared_ptr<File>` instead? So that they own the memory?

Comment (DarkDimius) [reply main/lsp/lsp.h]: (yeah, it does make me nervous)

File: main/lsp/lsp.h [hunk 6]
```diff
@@ -238,26 +251,34 @@ class LSPLoop {
...
+                                           const DidOpenTextDocumentParams &openParams) const;
+    TypecheckRun handleSorbetWorkspaceEdit(std::unique_ptr<core::GlobalState> gs,
+                                           const DidCloseTextDocumentParams &closeParams) const;
+    TypecheckRun handleSorbetWorkspaceEdit(std::unique_ptr<core::GlobalState> gs,
+                                           const WatchmanQueryResponse &queryResponse) const;
+    TypecheckRun handleSorbetWorkspaceEdits(std::unique_ptr<core::GlobalState> gs,
+                                            std::vector<std::unique_ptr<SorbetWorkspaceEdit>> &edits) const;
+    TypecheckRun commitSorbetWorkspaceEdits(std::unique_ptr<core::GlobalState> gs,
+                                            UnorderedMap<std::string, SorbetWorkspaceFileUpdate> &updates) const;
+    static std::string_view getFileContents(UnorderedMap<std::string, LSPLoop::SorbetWorkspaceFileUpdate> &updates,
```
Comment (DarkDimius) [inline main/lsp/lsp.h]: it seems that a bunch of helper methods below `hasSimilarName` & others can also be made `const`.

Comment (jvilk-stripe) [reply main/lsp/lsp.h]: ? this methid is `static`. `hasSimilarName` and friends aren't class methods, they are just in the namespace -- perhaps they should be static methods on the class instead.

Comment (DarkDimius) [reply main/lsp/lsp.h]: /me got confused. you're right

File: main/lsp/requests/sorbet_workspace_edit.cc [hunk 7]
```diff
@@ -128,9 +128,7 @@ LSPLoop::commitSorbetWorkspaceEdits(unique_ptr<core::GlobalState> gs,
         }
         return tryFastPath(move(gs), move(fileUpdates));
     } else {
-        TypecheckRun emptyRun;
-        emptyRun.gs = move(gs);
-        return emptyRun;
+        return TypecheckRun{{}, {}, {}, move(gs), {}, true};
```
Comment (DarkDimius) [inline main/lsp/requests/sorbet_workspace_edit.cc]: to prevent this landmine in the future, can we give all fields default initializers?
